### PR TITLE
Adding various build enhancements

### PR DESCRIPTION
**Pull request by MbolotSuse**
Adds the following enhancements:
- drone pipeline
- makefile for builds
- switches images to sle
- enables golangci in the validate stage
- changes various resources to use the cattle-csp-system namespace

### CODE REVIEW

**Review by ryansann — COMMENTED** [review]
It's not clear what changed in `pkg/supportconfig/handler.go` but otherwise looks good

**Comment by MbolotSuse** [comment]
> It's not clear what changed in pkg/supportconfig/handler.go but otherwise looks good

I think that golangci was complaining that it wasn't `go-fmt'ed` - though a diff didn't show any change. So IDK either, but this seems to pass `make validate-ci`.

**Comment by MbolotSuse** [comment]
@cmurphy Would appreciate your thoughts on `pkg/supportconfig/auth.go`. Feel free to review the rest of the PR if you want as well.

Auth Functionality:
- Safeguards the endpoint that allows you to get a support bundle (in case this contains sensitive info, we want to make sure only authorized users can access)
- Authentication requires that you submit a `Authorization: Bearer $TOKEN` header (where $TOKEN is a k8s token). It then uses a tokenReview to get the user's info and perform authentication.
- Authorization checks that you have access to "retrieve" on "supportconfigs.management.cattle.io" using a SAR and the user info retrieved during authentication.
  - This is not a real crd
  - The idea here is that only an admin or someone with specific permissions will be able to access this CRD (and therefore read this endpoint)
  - This is a bit sketchy since the CRD isn't real
  - I could swap this to a real configmap in this namespace that actually provides config values.

**Review by ryansann — APPROVED** [review]

File: .drone.yml [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    prerelease: true
+    checksum:
+      - sha256
+    checksum_file: CHECKSUMsum-amd64.txt
+    checksum_flatten: true
+    files:
+      - "dist/artifacts/*"
+  when:
+    ref:
+      - refs/head/master
```
Comment (cmurphy) [inline .drone.yml]: Could this be omitted? There is no master branch on this repo

File: pkg/clients/k8s/client.go [hunk 2]
```diff
@@ -5,12 +5,16 @@ import (
...
+	authnv1 "k8s.io/client-go/kubernetes/typed/authentication/v1"
+	authzv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
 	"k8s.io/client-go/rest"
 )
 
 type Clients struct {
 	ConfigMaps v1.ConfigMapClient
 	Secrets    v1.SecretController
+	SAR        authzv1.SubjectAccessReviewInterface
+	TR         authnv1.TokenReviewInterface
```
Comment (cmurphy) [inline pkg/clients/k8s/client.go]: Would be better to spell these out in full rather than abbreviate

```suggestion
	SubjectAccessReviews        authzv1.SubjectAccessReviewInterface
	TokenReviews         authnv1.TokenReviewInterface
```

File: pkg/supportconfig/auth.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	if !resp.Status.Allowed {
+		// if we aren't allowed, return an error with the reason
+		return fmt.Errorf("%s", resp.Status.Reason)
+	}
+	return nil
+}
+
+// getTokenFromHeader finds the converts the header "Authorization: Bearer MyTokenHere" -> MyTokenHere, and returns an
+// error if it fails to do so
+func getTokenFromHeader(r *http.Request) (string, error) {
```
Comment (cmurphy) [inline pkg/supportconfig/auth.go]: You could use this oneliner instead of this function: https://github.com/rancher/rancher/blob/84c32dde01e4e121c6aad80ef2cdd001fd5f7e55/pkg/api/steve/proxy/auth.go#L35

Or you could reuse this code if you want to allow basic auth and session cookies: https://github.com/rancher/rancher/blob/054b5d195f115b91a5284e85270403ba6f3ac99b/pkg/auth/tokens/token_util.go#L61

File: pkg/supportconfig/auth.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/sirupsen/logrus"
+	authv1 "k8s.io/api/authentication/v1"
+	v1 "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	authnv1 "k8s.io/client-go/kubernetes/typed/authentication/v1"
+	authzv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+)
+
+const (
+	unauthenticatedError = "unauthenticated \n"
```
Comment (cmurphy) [inline pkg/supportconfig/auth.go]: I think in most cases it's not a good practice to include formatting information (\n) in the error string. What's the context where this will be used, why is it not enough to just return a status code and have the UI handle the user-facing response page?

File: pkg/supportconfig/auth.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+func (a *AuthHandler) authorize(r *http.Request, userInfo *authv1.UserInfo) error {
+	//TODO: Real Resource?
+	resp, err := a.subjectReview.Create(r.Context(), &v1.SubjectAccessReview{
+		Spec: v1.SubjectAccessReviewSpec{
+			ResourceAttributes: &v1.ResourceAttributes{
+				Verb:     "retrieve",
+				Version:  "v1",
+				Resource: "supportconfigs",
+				Group:    "management.cattle.io",
+			},
```
Comment (cmurphy) [inline pkg/supportconfig/auth.go]: How did you test this? It's not possible to create a role for a resource type that doesn't exist

```
$ kubectl create role getsupportconfigs --verb=retrieve --resource=supportconfigs.management.cattle.io
Warning: 'retrieve' is not a standard resource verb
the server doesn't have a resource type "supportconfigs" in group "management.cattle.io"
```

I suggest using the endpoint path instead:

```suggestion
			NonResourceAttributes: &v1.NonResourceAttributes{
				Path: r.URL.Path,,
				Verb: r.Method,
			},
```

which could be used with a clusterrole like this:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: supportconfigs
rules:
- nonResourceURLs: ["/generate/supportconfig"]
  verbs: ["GET"]
```

File: pkg/supportconfig/auth.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+func (a *AuthHandler) authenticate(r *http.Request) (*authv1.UserInfo, error) {
+	token, err := getTokenFromHeader(r)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := a.tokenReview.Create(r.Context(), &authv1.TokenReview{
+		Spec: authv1.TokenReviewSpec{
+			Token: token,
+		},
+	}, metav1.CreateOptions{})
```
Comment (cmurphy) [inline pkg/supportconfig/auth.go]: I don't fully understand the context of how and where this application will be run so it's hard to evaluate this, but it looks to me like this is mainly useful for authenticating service accounts directly with the kubernetes cluster. Is that how this is meant to be used? Are real users supposed to be able to generate API tokens to use here? Does the dashboard need to be able to use this?

File: pkg/supportconfig/auth.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+	headerVal = strings.TrimSpace(headerVal)
+	splitHeader := strings.Split(headerVal, " ")
+	if len(splitHeader) != 2 {
+		return "", fmt.Errorf("header value for %s malformed", authHeader)
+	}
+	// TODO: Check value[0] and ensure it is "Bearer"?
+	return splitHeader[1], nil
+}
+
+// toExtra converts from one type of ExtraValue to another. I dislike this, but it does seem to be necessary
```
Comment (cmurphy) [inline pkg/supportconfig/auth.go]: This is a totally personal nitpicky opinion so feel free to ignore this feedback but I wouldn't include personal commentary in comments like this, it's not necessary and doesn't help someone in the future understand why it's there. If you feel like personal commentary is really important, include your name along with "I" statements because git blame can sometimes get messed up and it's not clear who "I" is who dislikes what's going on here.

fwiw I don't dislike this, it is indeed necessary because Go is a strongly typed language, and we have examples elsewhere doing the same thing https://github.com/rancher/rancher/blob/054b5d195f115b91a5284e85270403ba6f3ac99b/pkg/auth/webhook/webhook.go#L42-L47

File: pkg/server/server.go [hunk 8]
```diff
@@ -38,8 +41,9 @@ func setCertificateExpirationDays() error {
 }
 
 const (
-	port      = 8443
-	namespace = "cattle-system"
+	// TODO: Port from env var
+	port      = 9443
+	namespace = "cattle-csp-system"
```
Comment (cmurphy) [inline pkg/server/server.go]: Is there something somewhere that is going to create this namespace? I don't see it in the chart.

File: pkg/server/server.go [hunk 9]
```diff
@@ -24,6 +24,9 @@ func ListenAndServe(ctx context.Context, cfg *rest.Config, clients *k8s.Clients,
 	}
 
 	router := mux.NewRouter()
+	authHandler := supportconfig.NewAuthHandler(clients.SAR, clients.TR)
+	// handle auth before allowing support config to be created
```
Comment (cmurphy) [inline pkg/server/server.go]: (nit)

```suggestion
	// handle auth before allowing supportconfig to be created
```

A supportconfig is a suse thing, it's one word

**Review by cmurphy — CHANGES_REQUESTED** [review]
It looks like you accidentally did a git pull instead of a rebase, can you remove the merge commit:

```
git reset --hard HEAD^
git fetch origin
git rebase origin/release/v2.6
(fix conflicts)
```

It doesn't look like drone ran on this PR, can we get that set up before merging it? Or is that a chicken/egg situation?

One last request, could you possibly separate out the build and linter changes from the auth changes into separate commits since they are unrelated to each other?

File: pkg/supportconfig/auth.go (hunk 5 above)
Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: > How did you test this? It's not possible to create a role for a resource type that doesn't exist

Admin binding with a service account (works since admin has *). Technically, full admins are the only ones who need to be able to access this resource, though I really want to leave it open to be able to add/remove granularly in the future. 

One question on the endpoint path- My understanding is that nonResourceURLs are for k8s urls which deal with non-resource endpoints (like `$my_k8s_api_server/healthz`). In this case, we would be requiring that access is given to `$my_k8s_api_server/generate/supportconfig` in addition to the endpoint here (meaning that if this was ever an endpoint for k8s this would give more access than intended). 

Unlikely case, but I still think it's worth asking about. Do I have the correct understanding?

File: pkg/supportconfig/auth.go (hunk 4 above)
Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: I was testing this endpoint with curl and disliked the lack of feedback that the endpoint gave (had to use verbose mode to see the response code in the header). 

I can remove this and have it only set the header, I was just trying to make the response include some more info so that it was an easier-to-use api.

File: pkg/supportconfig/auth.go (hunk 7 above)
Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: > This is a totally personal nitpicky opinion so feel free to ignore this feedback but I wouldn't include personal commentary in comments like this, it's not necessary and doesn't help someone in the future understand why it's there.

This is fair and accurate feedback. I will remove the personal commentary. 

> fwiw I don't dislike this, it is indeed necessary because Go is a strongly typed language, and we have examples elsewhere doing the same thing

The thing that I disliked about it (FWIW) is that technically the ExtraValues are both just type aliases to `[]string`. So although they may appear to be a different type, they are really the same type, and I am wasting CPU cycles on converting between two things that are the same. But that's neither here nor there - appears to be necessary.

File: pkg/server/server.go (hunk 8 above)
Comment (MbolotSuse) [reply pkg/server/server.go]: I can create it in the chart if you want, but typically you can do this with `helm install --namespace cattle-csp-system --create-namespace`. FWIW, core rancher appears to take a similar approach, with the [docs](https://rancher.com/docs/rancher/v2.6/en/installation/install-rancher-on-k8s/#install-the-rancher-helm-chart) recommending that you manually create `cattle-system` before you install rancher.

That being said, I definitely see the use case of creating with the chart so if you feel strongly on this let me know.

File: pkg/supportconfig/auth.go (hunk 6 above)
Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: Apologies for the lack of context.

This is designed to run within EKS clusters running rancher 2.6.6 or higher. The functionality is to allow admin users to call this endpoint and generate a support bundle. In order to not re-implement significant auth functionality, I figured that we could just require users to supply the creds to this endpoint the same way they do to any other part of rancher. Since you can use tokens as kubeconfig tokens (and that's what the UI uses to get resources from rancher), I figured that the end result would be a link in the UI which redirects you to this service, and you supply the creds the same way you do to the rest of rancher. 

TL;DR - it's supposed to be used for getting a user from a supplied rancher api token (generated by the UI or user). This wasn't tested yet (I used an SA token when testing). Given your comment here, probably something that I should look at. Will post back here if it works.

**Comment by MbolotSuse** [comment]
@cmurphy I'll fix the commit log as requested. As far as drone goes, I should be able to get that working before merge. Will look into it and see if I can get something running.

File: pkg/supportconfig/auth.go (hunk 6 above)
Comment (cmurphy) [reply pkg/supportconfig/auth.go]: > This is designed to run within EKS clusters running rancher 2.6.6 or higher

Is the EKS cluster the local cluster or the downstream cluster?

> it's supposed to be used for getting a user from a supplied rancher api token

Yeah I couldn't get that to work, that token is a Rancher construct and doesn't go to k8s for validation

File: pkg/server/server.go (hunk 8 above)
Comment (cmurphy) [reply pkg/server/server.go]: Whichever way is consistent with how it is being installed is fine. If users are supposed to use the helm CLI directly then `--create-namespace` is fine. If they are meant to install it through the Rancher apps page then it will need to work with that system.

File: pkg/supportconfig/auth.go (hunk 5 above)
Comment (cmurphy) [reply pkg/supportconfig/auth.go]: That's a good point, using nonResourceURLs here is still somewhat abusing k8s RBAC. If all we need for now is to check whether the user is an admin, could we just check for '*' directly? 
```suggestion
			ResourceAttributes: &v1.ResourceAttributes{
				Verb:     "*",
				Version:  "",
				Resource: "*",
				Group:    "",
			},
```

File: pkg/supportconfig/auth.go (hunk 6 above)
Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: > Is the EKS cluster the local cluster or the downstream cluster?

Local cluster. 

> Yeah I couldn't get that to work, that token is a Rancher construct and doesn't go to k8s for validation

Ah. I should have tested that scenario, sorry about that. One thing to keep in mind - this won't go directly from the user/browser to the service. The request will need to be proxied through rancher (i.e. /k8s/local/ was how it was described to me). I'm thinking that due to impersonation the end creds that will be used will be an SA token, but that's just a guess. I'll test this out and get a better approach together that is more realistic.

Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: Ok - ran some tests and have some more info. 

The idea is that when the user clicks on "give me a supportconfig" button in the UI, it will redirect them to the service proxy URL for the service (`$RANCHER_URL/api/v1/namespaces/cattle-csp-system/services/https:rancher-csp-adapter:https/proxy/generate/supportconfig`). This does not appear to pass along the header like I thought. I will see what options I have here - maybe there is a header that I can use to do what I need.

File: pkg/supportconfig/auth.go (hunk 5 above)
Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: The issue there is that this leaves no room for people to use a custom role - the user that hits this endpoint must be effectively an admin user and nothing else. I'd really like to leave this open to be used in a customized manner. 

I'm kind of thinking that your original idea of a non-resource-url might work as long as we give a prefix that would never be part of the k8s api (like `/rancher-csp-adapter/generate/supportconfig` or maybe that w/o the forward slash).

File: pkg/clients/k8s/client.go (hunk 2 above)
Comment (MbolotSuse) [reply pkg/clients/k8s/client.go]: Resolving - support config functionality and all related auth was removed from this PR.

File: pkg/supportconfig/auth.go (hunk 3 above)
Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: Resolving due to change of goals with this PR.

File: pkg/supportconfig/auth.go (hunk 4 above)
Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: Resolving due to change of goals with this PR.

File: pkg/supportconfig/auth.go (hunk 5 above)
Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: Resolving due to change of goals with this PR.

File: pkg/supportconfig/auth.go (hunk 7 above)
Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: Resolving due to change of goals with this PR.

File: pkg/server/server.go (hunk 8 above)
Comment (MbolotSuse) [reply pkg/server/server.go]: Resolving due to change in goals with this PR.

File: pkg/server/server.go (hunk 9 above)
Comment (MbolotSuse) [reply pkg/server/server.go]: Resolving due to change in goals with this PR.

File: pkg/supportconfig/auth.go (hunk 6 above)
Comment (MbolotSuse) [reply pkg/supportconfig/auth.go]: Resolving - supportconfig functionality will not be altered here at this time.

File: .drone.yml (hunk 1 above)
Comment (MbolotSuse) [reply .drone.yml]: I have removed this from the amd and arm stages

File: pkg/server/server.go (hunk 8 above)
Comment (MbolotSuse) [reply pkg/server/server.go]: I have left this as is - I think that having the users specify those options on install (or create the namespace beforehand) mirrors what rancher does, and does fit within acceptable conventions. If you disagree, please let me know.

**Comment by MbolotSuse** [comment]
Made some changes to the PR and the goals of this PR. There are no more adjustments to the supportconfig or auth generations. Instead, this is just getting the pipeline setup, creating a makefile, and changing to sle bci.

**Comment by MbolotSuse** [comment]
Made some changes to the PR and the goals of this PR. There are no more adjustments to the supportconfig or auth generations. Instead, this is just getting the pipeline setup, changing to sle bci, and fixing bugs with make and it's related tooling.

**Review by ryansann — APPROVED** [review]

**Review by cmurphy — APPROVED** [review]
